docs/knowledgebase/runtime/macros.md

```diff
@@ -2,58 +2,270 @@
 title: Runtime Macros
 ---
 
-The Substrate runtime macros are a set of utilities which make the development of runtime modules
-simpler for you.
+## Introduction
 
-## What are Rust macros?
+Rust macro is a way of writing code that writes code, also called metaprogramming. Generally speaking it can be designed to read, generate or transform other programs. Thus allowing programmers to minimize the number of lines of code to express a solution. It also allows programs greater flexibility to efficiently handle new situations without recompilation. Macro helps create domain specific language (DSL) by abstracting out some implementation details and defining its own rules and logics.
 
-In short, Rust macros are a way to write code that writes more code. Macros work by matching against
-the specific patterns defined in the macro rule, capturing part of the match as variables, and then
-expand to produce even more code.
+In Substrate runtime development context, this helps runtime engineers to focus on the runtime logics and forget about details such as how the on-chain variables need to be encoded and decoded. That's why developers frequently leverage these macros to simplify runtime development effort once the intent is made known via the succinct statements in macros. But the downside is that sometimes developers may encounter error messages of new structs or types that seem to come out of nowhere, but actually as a result of how the macros create new types and interact with contents inside.
```

Comment:
I would delete the entire first paragraph and start with the second one. This article is about how Substrate uses macros, not about what Rust macros are.

Reply:
rewrote first section